### PR TITLE
Lower _d_arraycast to object.__ArrayCast

### PR DESCRIPTION
**Pull request by JinShil**
Requires https://github.com/dlang/druntime/pull/2264 to be merged first

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -6383,6 +6397,42 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+            {
+                // Array element sizes do not match, so we must adjust the dimensions
+
+                if ((fromSize % toSize) != 0)
+                {
+                    // Runtime check needed in case arrays don't line up
+
+                    // lower to `object.__ArrayCast!(TFrom, TTo)(from)`
+
+                    auto id = Identifier.idPool("__ArrayCast");
```
Comment (JinShil) [inline src/dmd/expressionsem.d]: How can I fully qualify this to `object.__ArrayCast`?

File: test/runnable/betterc_array_cast.d [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    // We can't actually create dynamic arrays in idiomatic D when
+    // compiling with -betterC, so we do it manually.
+    auto b_length = cast(size_t*)&b;
+    auto b_ptr = cast(void*)(b_length + 1);
+    *b_length = int.sizeof * 3;
+    b_ptr = malloc(*b_length);
+
+    // size mismatch, should result in an assertion failure
+    l = cast(long[])b;
+}
```
Comment (Geod24) [inline test/runnable/betterc_array_cast.d]: Missing newline

Comment (JinShil) [reply test/runnable/betterc_array_cast.d]: Fixed.

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (jacob-carlborg) [reply src/dmd/expressionsem.d]: Look at how the comparison lowering is done to `__cmp`.

https://github.com/dlang/dmd/blob/03e7693150c453055b098768e59ebd31cb3ca52e/src/dmd/expressionsem.d#L9794-L9798

Comment (JinShil) [reply src/dmd/expressionsem.d]: But I need an `Identifer`, not an `Expression`, for the `TemplateInstance`, later.  How do I create a fully-qualified `Identifier`?

Comment (jacob-carlborg) [reply src/dmd/expressionsem.d]: > How do I create a fully-qualified Identifier?

I don't think that is possible. To get the fully qualified name inside the compiler one needs to walk up the parent chain and build up the name that way. In your case, I think you need to use the overload of the `TemplateInstance` constructor that takes a `TemplateDeclaration`. Perhaps you can store the `TemplateDeclaration` when it's analyzed in the same way as the `Object` class declaration (and others) is stored:

https://github.com/dlang/dmd/blob/03e7693150c453055b098768e59ebd31cb3ca52e/src/dmd/dclass.d#L367-L372

**Pull request closed without merging**